File: files/2025-04-06/CSEC 06042025.R
library(tidyverse)
library(readxl)

path = "files/Challenge1425.xlsx"
input = read_excel(path, range = "B3:D23")
test  = read_excel(path, range = "F3:H7")

input = input %>%
  mutate(Date = str_replace(Date, "29-02", "27-02") %>% dmy())

seq = seq.Date(min(input$Date), max(input$Date), by = "1 day") %>%
  data.frame(Date = .) %>%
  left_join(input, by = "Date") %>%
  fill(`Staff No.`, .direction = "down") %>%
  mutate(week = isoweek(Date)) %>%
  summarise(`Total Hours` = sum(`Worked Hours`, na.rm = TRUE),
            `Week Dates` = paste(min(Date), max(Date), sep = " - "),
            .by = c("Staff No.", "week")) %>%
  select(1,4,3)

File: files/2025-04-06/CSEC 06042025.py
import pandas as pd
import numpy as np

path = "files/Challenge1425.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=2, nrows=21)
test = pd.read_excel(path, usecols="F:H", skiprows=2, nrows=4)

input['Date'] = input['Date'].str.replace("29-02", "27-02", regex=False)
input['Date'] = pd.to_datetime(input['Date'], format='%d-%m-%Y')

date_range = pd.date_range(start=input['Date'].min(), end=input['Date'].max(), freq='D')
seq = pd.DataFrame({'Date': date_range})

seq = seq.merge(input, on='Date', how='left')
seq['Staff No.'] = seq['Staff No.'].ffill()

seq['week'] = seq['Date'].dt.isocalendar().week

summary = (
    seq.groupby(['Staff No.', 'week'])
    .agg(
        **{
            'Total Hours': ('Worked Hours', lambda x: x.sum(skipna=True)),
            'Week Dates': ('Date', lambda x: f"{x.min().strftime('%Y-%m-%d')} - {x.max().strftime('%Y-%m-%d')}")
        }
    )
    .reset_index()
)
summary = summary[['Staff No.', 'Week Dates', 'Total Hours']]